### PR TITLE
Shell layout persistence improvements

### PR DESCRIPTION
**Pull request by vzhukovs**
### What does this PR do?
This improvement proposal provides refactored mechanism for shell layout persistence. 

Highlighted items extracted from the current proposal change:

- Shell layout save timeout changed from 10 seconds to 2 seconds;
- Removed server side implementation that operated with the storage files and used [FileService](https://github.com/eclipse-theia/theia/blob/42042974c59f148974de5e8fd90cba7222cee82f/packages/filesystem/src/browser/file-service.ts#L271) instead;
- Created an implementation of [FrontendApplicationContribution](https://github.com/eclipse-theia/theia/blob/42042974c59f148974de5e8fd90cba7222cee82f/packages/core/src/browser/frontend-application.ts#L33) to allow Che-Theia subscribe to additional window and document events to control UI activity and persist state;
- Created an extension of [ShellLayoutRestorer](https://github.com/eclipse-theia/theia/blob/42042974c59f148974de5e8fd90cba7222cee82f/packages/core/src/browser/shell/shell-layout-restorer.ts#L119) to control the persistence, when Theia goes into `closing_window` state. In this state performing the requests through the websocket can be unexpected, because in this case Theia tries to shut down the registered contributions and websocket channels might be closed;
- Created caching mechanism to prevent storing the same JSON object which was serialized from the application state;
- Removed `che-storage-frontend-contribution.ts` as this implementation implemented in `CheFrontendApplication`;
- Changed event subscription from monitoring add/remove/active widgets to keyboard/mouse events, because this type of events can handle cursor changing position inside opened editor(implementation in master doesn't store cursor position);
- Align with the vanilla Theia store mechanism to persist shell layout across opened workspace roots.

This comes from the provided investigation: https://github.com/eclipse/che/issues/18564#issuecomment-766182346

Signed-off-by: Vladyslav Zhukovskyi <vzhukovs@redhat.com>

### Screenshot/screencast of this PR
State in master branch:
pannels-folded

State within current PR's implementation:
pannels-restored

### What issues does this PR fix or reference?
https://github.com/eclipse/che/issues/18564

### How to test this PR?
Use the following devfile to create the workspace on che.openshift.io:
```yaml
apiVersion: 1.0.0
metadata:
  name: wksp-restore
attributes:
  persistVolumes: 'false'
projects:
  - name: console-java-simple
    source:
      location: 'https://github.com/che-samples/console-java-simple.git'
      type: git
      branch: master
components:
  - type: cheEditor
    reference: 'https://gist.github.com/vzhukovs/732676571880f51bc0b87cc2f7ee1304/raw/c5f007672a49416d66239882da28463f77f96e3b/meta.yaml'
```

After opening the workspace try to wait for 2 seconds and refresh the browser page. State should be restored after page reload.

### PR Checklist

[As the author of this Pull Request I made sure that:](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#pull-request-template-and-its-checklist)

- [x] [The Eclipse Contributor Agreement is valid](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#the-eclipse-contributor-agreement-is-valid)
- [x] [Code produced is complete](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#code-produced-is-complete)
- [x] [Code builds without errors](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#code-builds-without-errors)
- [ ] [Tests are covering the bugfix](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#tests-are-covering-the-bugfix)
- [x] [The repository devfile is up to date and works](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#the-repository-devfile-is-up-to-date-and-works)
- [x] [Sections `What issues does this PR fix or reference` and `How to test this PR` completed](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#sections-what-issues-does-this-pr-fix-or-reference-and-how-to-test-this-pr-completed)
- [ ] [Relevant user documentation updated](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#relevant-contributing-documentation-updated)
- [ ] [Relevant contributing documentation updated](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#relevant-contributing-documentation-updated)
- [ ] [CI/CD changes implemented, documented and communicated](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#cicd-changes-implemented-documented-and-communicated)

### Reviewers

Reviewers, please comment how you tested the PR when approving it.

### Happy Path Channel

HAPPY_PATH_CHANNEL=stable

### CODE REVIEW

File: extensions/eclipse-che-theia-user-preferences/src/browser/che-frontend-application.ts [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+    super(commands, menus, keybindings, layoutRestorer, contributions, _shell, stateService);
+  }
+
+  protected registerEventListeners() {
+    super.registerEventListeners();
+
+    window.addEventListener('resize', () => this.scheduleLayoutPersistence());
+    document.addEventListener('keydown', () => this.scheduleLayoutPersistence());
+    document.addEventListener('mousedown', () => this.scheduleLayoutPersistence());
+    document.addEventListener('mousemove', () => this.scheduleLayoutPersistence());
```
Comment (RomanNikitenko) [inline extensions/eclipse-che-theia-user-preferences/src/browser/che-frontend-application.ts:49]: Do we really need to track `mousemove` event?
I mean - what info we are going to persist at such event and if it's possible to replace the event by more target one.

Comment (vzhukovs) [reply extensions/eclipse-che-theia-user-preferences/src/browser/che-frontend-application.ts:49]: Possible drag and drop operations, such as moving the editor tabs.

Comment (RomanNikitenko) [reply extensions/eclipse-che-theia-user-preferences/src/browser/che-frontend-application.ts:49]: I asked because the event will request required operation too often as for me.
For example, I did a few clicks during couple of seconds and got hundreds such events

onMouseEvent

File: extensions/eclipse-che-theia-user-preferences/src/browser/che-frontend-application.ts [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+  }
+
+  protected scheduleLayoutPersistence() {
+    if (this.layoutPersistenceTimer) {
+      this.clearTimeout(this.layoutPersistenceTimer);
+    }
+
+    this.layoutPersistenceTimer = this.setTimeout(() => {
+      this.layoutRestorer.storeLayout(this);
+    }, this.storagePreferences['workbench.layout.saveTimeout']);
```
Comment (RomanNikitenko) [inline extensions/eclipse-che-theia-user-preferences/src/browser/che-frontend-application.ts:59]: should we set field `this.layoutPersistenceTimer` to `undefined` here?

File: extensions/eclipse-che-theia-user-preferences/src/browser/che-frontend-application.ts [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+    super.registerEventListeners();
+
+    window.addEventListener('resize', () => this.scheduleLayoutPersistence());
+    document.addEventListener('keydown', () => this.scheduleLayoutPersistence());
+    document.addEventListener('mousedown', () => this.scheduleLayoutPersistence());
+    document.addEventListener('mousemove', () => this.scheduleLayoutPersistence());
+  }
+
+  protected scheduleLayoutPersistence() {
+    if (this.layoutPersistenceTimer) {
```
Comment (RomanNikitenko) [inline extensions/eclipse-che-theia-user-preferences/src/browser/che-frontend-application.ts:53]: looks like it's always `true` except the first method call

File: extensions/eclipse-che-theia-user-preferences/src/browser/che-frontend-application.ts (hunk 1 above)
Comment (vzhukovs) [reply extensions/eclipse-che-theia-user-preferences/src/browser/che-frontend-application.ts:49]: It is expected. Trigger for layout persistence will be activated after 2 seconds, when mouse stops.
We have similar implementation for activity tracker extension: [che-theia-activity-tracker-contribution.ts#L38](https://github.com/eclipse/che-theia/blob/master/extensions/eclipse-che-theia-activity-tracker/src/browser/che-theia-activity-tracker-contribution.ts#L38)

Moreover, this is common practice in js environment: 
https://stackoverflow.com/a/610000
https://stackoverflow.com/a/19262511
https://stackoverflow.com/a/15067924

File: extensions/eclipse-che-theia-user-preferences/src/browser/che-frontend-application.ts (hunk 2 above)
Comment (vzhukovs) [reply extensions/eclipse-che-theia-user-preferences/src/browser/che-frontend-application.ts:59]: Does not necessarily. From the documentation:
>Passing an invalid ID to clearTimeout() silently does nothing; no exception is thrown.

The same we can observe in [FrontendConnectionStatusService](https://github.com/eclipse-theia/theia/blob/42042974c59f148974de5e8fd90cba7222cee82f/packages/core/src/browser/connection-status-service.ts#L157-L171).

File: extensions/eclipse-che-theia-user-preferences/src/browser/che-frontend-application.ts (hunk 3 above)
Comment (vzhukovs) [reply extensions/eclipse-che-theia-user-preferences/src/browser/che-frontend-application.ts:53]: It is true. Answered in https://github.com/eclipse/che-theia/pull/970#discussion_r563931483

File: extensions/eclipse-che-theia-user-preferences/src/browser/che-frontend-application.ts (hunk 1 above)
Comment (RomanNikitenko) [reply extensions/eclipse-che-theia-user-preferences/src/browser/che-frontend-application.ts:49]: AFAIK the activity tracker sends request once per 1 minute, but it can be 30 times per 1 minute.
But OK if we don't have another way to detect drag and drop operations.

Comment (vzhukovs) [reply extensions/eclipse-che-theia-user-preferences/src/browser/che-frontend-application.ts:49]: We can track drag event with implementation that listens to drag enter/leave/drop events, but this mechanism will be redundant for this tiny functionality.

File: extensions/eclipse-che-theia-user-preferences/src/browser/che-frontend-application.ts (hunk 3 above)
Comment (RomanNikitenko) [reply extensions/eclipse-che-theia-user-preferences/src/browser/che-frontend-application.ts:53]: so we can remove it at all? 
:-)

**Review by azatsarynnyy — COMMENTED** [review]
I've tested the proposed patch yesterday, on both Hosted Che and [Developer Sandbox](https://developers.redhat.com/developer-sandbox).
The issue can't be reproduced anymore. Great job!
I plan to look at the code a bit later today.

File: extensions/eclipse-che-theia-user-preferences/src/browser/che-frontend-application.ts [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+} from '@theia/core/lib/browser';
+import { CommandRegistry, ContributionProvider, MenuModelRegistry } from '@theia/core';
+import { inject, injectable, named } from 'inversify';
+
+import { CheShellLayoutRestorer } from './che-shell-layout-restorer';
+import { FrontendApplicationStateService } from '@theia/core/lib/browser/frontend-application-state';
+import { StorageServicePreferences } from './che-storage-preferences';
+
+@injectable()
+export class CheFrontendApplication extends FrontendApplication {
```
Comment (azatsarynnyy) [inline extensions/eclipse-che-theia-user-preferences/src/browser/che-frontend-application.ts:25]: Is there any good reason why we need to rebind the upstream component and depend on it?
Can we register the required listeners by providing [`FrontendApplicationContribution`](https://github.com/eclipse-theia/theia/blob/42042974c59f148974de5e8fd90cba7222cee82f/packages/core/src/browser/frontend-application.ts#L33)?

File: extensions/eclipse-che-theia-user-preferences/src/browser/che-shell-layout-restorer.ts [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+import {
+  FrontendApplicationState,
+  FrontendApplicationStateService,
+} from '@theia/core/lib/browser/frontend-application-state';
+import { inject, injectable } from 'inversify';
+
+import { ILogger } from '@theia/core';
+
+@injectable()
+export class CheShellLayoutRestorer extends ShellLayoutRestorer {
```
Comment (azatsarynnyy) [inline extensions/eclipse-che-theia-user-preferences/src/browser/che-shell-layout-restorer.ts:20]: Let's document all the overridden functionality here.
Because it's not obvious why we're patching it downstream instead of providing it upstream.
Several comments will help a lot when we'll need to look at this code in a year.

File: extensions/eclipse-che-theia-user-preferences/src/browser/che-frontend-application.ts (hunk 4 above)
Comment (vzhukovs) [reply extensions/eclipse-che-theia-user-preferences/src/browser/che-frontend-application.ts:25]: Theoretically it can be realized through the `onDidInitializeLayout` implementation. Need to try.

File: extensions/eclipse-che-theia-user-preferences/src/browser/che-frontend-application.ts (hunk 3 above)
Comment (vzhukovs) [reply extensions/eclipse-che-theia-user-preferences/src/browser/che-frontend-application.ts]: Removed

File: extensions/eclipse-che-theia-user-preferences/src/browser/che-frontend-application.ts (hunk 4 above)
Comment (vzhukovs) [reply extensions/eclipse-che-theia-user-preferences/src/browser/che-frontend-application.ts]: Reimplemented

File: extensions/eclipse-che-theia-user-preferences/src/browser/che-shell-layout-restorer.ts (hunk 5 above)
Comment (vzhukovs) [reply extensions/eclipse-che-theia-user-preferences/src/browser/che-shell-layout-restorer.ts:28]: Added documentation section.

File: extensions/eclipse-che-theia-user-preferences/src/browser/che-frontend-application.ts (hunk 4 above)
Comment (azatsarynnyy) [reply extensions/eclipse-che-theia-user-preferences/src/browser/che-frontend-application.ts]: Thanks!

File: extensions/eclipse-che-theia-user-preferences/src/browser/che-shell-layout-restorer.ts (hunk 5 above)
Comment (azatsarynnyy) [reply extensions/eclipse-che-theia-user-preferences/src/browser/che-shell-layout-restorer.ts:28]: Thanks!

**Review by azatsarynnyy — APPROVED** [review]
Tested it earlier https://github.com/eclipse/che-theia/pull/970#pullrequestreview-576043405

**Review by RomanNikitenko — APPROVED** [review]
Tested on `minikube` and `dogfooding` instances.
Restoration of state works well for me:
- after refreshing a page
- after restarting a workspace

**Comment by azatsarynnyy** [comment]
> 18:50:35 error parsing HTTP 429 response body: invalid character 'T' looking for beginning of value: "Too Many Requests (HAP429).\n"

I'm rerunning the tests

**Comment by azatsarynnyy** [comment]
[crw-ci-test --rebuild]

**Comment by vzhukovs** [comment]
[crw-ci-test]

**Comment by vzhukovs** [comment]
[crw-ci-test --rebuild]

**Comment by vzhukovs** [comment]
[crw-ci-test]

**Comment by vzhukovs** [comment]
[crw-ci-test]

**Comment by dmytro-ndp** [comment]
[crw-ci-test]

File: extensions/eclipse-che-theia-user-preferences/src/browser/che-shell-layout-restorer.ts [hunk 6]
```diff
@@ -0,0 +1,63 @@
...
+import {
+  FrontendApplicationState,
+  FrontendApplicationStateService,
+} from '@theia/core/lib/browser/frontend-application-state';
+import { inject, injectable } from 'inversify';
+
+import { ILogger } from '@theia/core';
+
+/**
+ * Rebounded extension of ShellLayoutRestorer.
```
Comment (tsmaeder) [inline extensions/eclipse-che-theia-user-preferences/src/browser/che-shell-layout-restorer.ts:20]: Probable you mean "rebound"?

File: extensions/eclipse-che-theia-user-preferences/src/browser/che-shell-layout-restorer.ts [hunk 7]
```diff
@@ -0,0 +1,63 @@
...
+ * Extends super class by adding an ability to cache the application state which prevents redundant saving state
+ * if cached value is the same as incoming one. Modifies default behavior by disabling ability to store application
```
Comment (tsmaeder) [inline extensions/eclipse-che-theia-user-preferences/src/browser/che-shell-layout-restorer.ts:22]: ...if _the_ cached value... same as _the_ incoming one. Modifies _the_ default behavior by disabling _the_ ability

File: extensions/eclipse-che-theia-user-preferences/src/browser/che-shell-layout-restorer.ts [hunk 8]
```diff
@@ -0,0 +1,63 @@
...
+ * state if Theia is in closing_window state. In this state UI, which is serializing for the persistence is unexpected
+ * due to internal modifications DOM object. Some of widgets might have incorrect values for their size, visibility,
```
Comment (tsmaeder) [inline extensions/eclipse-che-theia-user-preferences/src/browser/che-shell-layout-restorer.ts:24]: > In this state UI, which is serializing for the persistence is unexpected due to internal modifications DOM object

I don't understand this sentence, could you rephrase?

**Review by tsmaeder — COMMENTED** [review]
Not trying to be nitpicky here...I guess articles are not used in Ukranian (as in Latin, btw.), but in English, omitting them makes the text hard to read.

**Comment by tsmaeder** [comment]
@vzhukovs 
> Created an extension of ShellLayoutRestorer to control the persistence, when Theia goes into closing_window state. In this state performing the requests through the websocket can be unexpected, because in this case Theia tries to shut down the registered contributions and websocket channels might be closed;

Does this not apply to upstream Theia as well? If not, what's the difference?

**Comment by vzhukovs** [comment]
@tsmaeder I've updated the doc section. 

> Does this not apply to upstream Theia as well? If not, what's the difference?

The main point is that vanilla Theia and Che-Theia use different types of implementations of StorageService, the first one is using the browser LocalStorage that is a priori faster that StorageService in Che-Theia that uses file system.

File: extensions/eclipse-che-theia-user-preferences/src/browser/che-shell-layout-restorer.ts [hunk 9]
```diff
@@ -0,0 +1,69 @@
...
+import {
+  FrontendApplicationState,
+  FrontendApplicationStateService,
+} from '@theia/core/lib/browser/frontend-application-state';
+import { inject, injectable } from 'inversify';
+
+import { ILogger } from '@theia/core';
+
+/**
+ * An extension of default ShellLayoutRestorer. Adds and ability to cache application state
```
Comment (tsmaeder) [inline extensions/eclipse-che-theia-user-preferences/src/browser/che-shell-layout-restorer.ts:20]: Fantastic explanation. Thanks and more of the same, please :-)

**Comment by tsmaeder** [comment]
> @tsmaeder I've updated the doc section.

Thx, @vzhukovs . The new explanation is exactly what I like to see in a comment: it describes the _intent_ of the code instead of just repeating what's already visible in the code. Kudos!
